Fix transform incorrectly calculating date bucket on updating old data

alyokaz opened the pull request:
This pull request fixes #97101, where a continuous pivot transform was incorrectly calculating a metric aggregation on a date bucket when a document from that bucket was updated. This was due to the transform taking the updated document's timestamp field as an upper bound in the transforms optimisation query and, consequently, ignoring all documents with a later timestamp in the updated metric aggregation calculation for the bucket.

I changed the the rounding of `upperBound` in `DateHistogramFieldCollector` to round up to the next time interval, as this is where the transform's optimisation query pulls its upper bound from. This will have the effect of all the documents in the same bucket as the document being updated being pulled into the re-calculation of the buckets aggregations.

I've updated and refactored out the unit tests from `CompositeBucketsChangeCollectorTests` into a new class `DateHistorgramFieldCollectorTests`. This was done in the process of reading and understanding the tests. I'm not sure if the refactor is necessary or adds unnecessary bloat, but the same could be done for the other `FieldCollector` tests. It has been added as a separate commit as advised. 

I've added an integration test in `TransformPivotRestIT`. I chose this rather then `TransformContinuousIT` as the specific set up needed to recreate the issue didn't seem to fit well with the approach used there. The test could be refactored / integrated into the test harness further, however, I thought I'd leave it as is just to be more explicit about what was done.

alyokaz committed:
Refactor CompositeBucketsChangeCollectorTests

Refactor tests for DateHistogramFieldCollector into multiple smaller tests
and move to separate file to increase clarity and extensibility.

alyokaz committed:
Fix transform incorrectly calculating date bucket on updating old data (#97101)

Fix transform ignoring docs in the same time bucket with a later date than
the updated doc when a doc in the time bucket is updated.

elasticsearchmachine commented:
Pinging @elastic/ml-core (Team:ML)

hendrikmuhs commented on x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/pivot/DateHistorgramFieldCollectorTests.java:
```diff
@@ -0,0 +1,163 @@
+    final double MAX_TIMESTAMP_VALUE = 302_525;
+    final double EXPECTED_LOWER_BOUND = 120_000;
+
+    final double EXPECTED_UPPER_BOUND = 360_000;
```
constants should be `private static final`

hendrikmuhs commented on x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/pivot/DateHistorgramFieldCollectorTests.java:
```diff
@@ -0,0 +1,163 @@
+    Map<String, SingleGroupSource> groups = new LinkedHashMap<>();
+
+    SingleValue minTimestamp = mock(NumericMetricsAggregation.SingleValue.class);
+    SingleValue maxTimestamp = mock(NumericMetricsAggregation.SingleValue.class);
```
a mix of per suite and per test (`@Before`) initialization is unclean. Apart from that, members should be `private`. I think it is cleaner if every test is independent and uses own mocks. You can still re-use code. My suggestion is to introduce but not initialize the variable:

```
private SingleValue minTimestamp;
private SingleValue maxTimestamp;
```

and set it up in 

```
    @Before
    public void setupDateHistogramFieldCollectorTest() {
         minTimestamp = mock(NumericMetricsAggregation.SingleValue.class);
         maxTimestamp = mock(NumericMetricsAggregation.SingleValue.class);
        ...
```

hendrikmuhs commented on x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/pivot/DateHistorgramFieldCollectorTests.java:
```diff
@@ -0,0 +1,163 @@
+    final double EXPECTED_UPPER_BOUND = 360_000;
+
+    @Before
+    public void setUpDateHistogramFieldCollectorTests() {
```
`setUp` -> `setup` (one word). `@Before` works per test, so `setupDateHistogramFieldCollectorTest` not `setupDateHistogramFieldCollectorTests`

hendrikmuhs commented on x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/pivot/DateHistorgramFieldCollectorTests.java:
```diff
@@ -0,0 +1,163 @@
+        assertNull(queryBuilder);
+    }
+
+    private void assertQuery(QueryBuilder queryBuilder, Double expectedLowerBound, Double expectedUpperBound, String expectedFieldName) {
```
this and the below owns should be static I think

hendrikmuhs commented:
@elasticmachine test this please

hendrikmuhs reviewed (commented):
Thanks for your contribution, I've triggered test execution in CI and did a 1st quick review and added some comments.

alyokaz committed:
Impelment Code Review Suggestions For DateHistogramFieldCollectorTests

Clean up test initialisation. Make member fields private and constants static.
Make utility methods/ pure functions static. Fix typos.

alyokaz commented:
I've made the constants private and static, moved the initialisation of the `minTimestamp`, `maxTimestamp` and `groups` to `setupDateHistogramFieldCollectorTest`, made the utility / pure functions static and fixed the typos as suggested.

hendrikmuhs commented:
@elasticmachine test this please

hendrikmuhs commented:
Please address the checkstyle issue, you can `gradlew precommit` locally, the output tells you how to format the code easily with another gradle command.

Please also merge main into the branch(feel free to use `rebase` to keep your commits on top).

The PR looks almost finished, good work!

alyokaz committed:
Fix Checkstyle Violations

Optimise imports. Reformat code.

alyokaz commented:
I've fixed the checkstyle violations and rebased the branch on main.

Thanks, glad it's up to scratch.

hendrikmuhs commented:
@elasticsearchmachine generate changelog

hendrikmuhs commented:
@elasticmachine test this please

hendrikmuhs commented:
@elasticmachine generate changelog

Commit:
add changelog

hendrikmuhs commented:
@elasticmachine test this please

hendrikmuhs reviewed (approved):
LGTM

Looks great! Thank you for your contribution.

alyokaz commented:
Happy to contribute, thanks for the review.
